src/main/scala/tech/cryptonomic/conseil/Lorre.scala

```diff
@@ -1,771 +1,53 @@
-        val processed = Some(ids.keySet)
-        logger.info("Cleaning {} processed accounts from the checkpoint...", ids.size)
-        db.run(TezosDb.cleanAccountsCheckpoint(processed))
-          .map(cleaned => logger.info("Done cleaning {} accounts checkpoint rows.", cleaned))
-      }
-
-      //if needed, we get the stored levels and only keep updates that are more recent
-      def prunedUpdates(): Future[Map[AccountId, BlockReference]] =
-        if (onlyProcessLatest) db.run {
-          TezosDb.getLevelsForAccounts(ids.keySet).map { currentlyStored =>
-            ids.filterNot {
-              case (AccountId(id), (_, updateLevel, _, _)) =>
-                currentlyStored.exists { case (storedId, storedLevel) => storedId == id && storedLevel > updateLevel }
-            }
-          }
-        } else Future.successful(ids)
-
-      logger.info("Ready to fetch updated accounts information from the chain")
-
-      /* Streams the (unevaluated) incoming data, actually fetching the results.
-       * We use combinators to keep the ongoing requests' flow under control, taking advantage of
-       * akka-streams automatic backpressure control.
-       * The results are grouped to optimize for database storage.
-       * We do this to re-aggregate results from pages which are now based on single blocks,
-       * which would lead to inefficient storage performances as-is.
-       */
-      val saveAccounts = (pages: LazyPages[tezosNodeOperator.AccountFetchingResults]) =>
-        Source
-          .fromIterator(() => pages)
-          .mapAsync(1)(identity) //extracts the future value as an element of the stream
-          .mapConcat(identity) //concatenates the list of values as single-valued elements in the stream
-          .grouped(batchingConf.blockPageSize) //re-arranges the process batching
-          .map(extractDelegatesInfo)
-          .mapAsync(1)((processAccountsPage _).tupled)
-          .runFold(Monoid[(Int, Option[Int], List[BlockTagged[DelegateKeys]])].empty) { (processed, justDone) =>
-            processed |+| justDone
-          } andThen logOutcome
-
-      val fetchAndStore = for {
-        (accountPages, _) <- prunedUpdates().map(tezosNodeOperator.getAccountsForBlocks)
-        (stored, checkpoints, delegateKeys) <- saveAccounts(accountPages) flatTap cleanup
-      } yield delegateKeys
-
-      fetchAndStore.transform(
-        identity,
-        e => {
-          val error = "I failed to fetch accounts from client and update them"
-          logger.error(error, e)
-          AccountsProcessingFailed(message = error, e)
-        }
-      )
-    }
-  }
-
-  private object DelegatesProcessor {
-
-    /* Fetches the data from the chain node and stores delegates into the data store.
-     * @param ids a pre-filtered map of delegate hashes with latest block referring to them
+    val indexingLoop: Kleisli[IO, ConfiguredResources, Unit] = Kleisli(
+      (LorreIndexing.run _).tupled
+    ).local(buildDeps)
```

Comment:
Notes: here we can see that in the end, the main loop is defined elsewhere and will need both general and specialized lorre's resources and operations to work (that's why we needed to compose those with the above `buildDeps` function.

The resulting Kleisli built from `LorreIndexing.run` will take both `ConfiguredResources` and `CompositeOps`, but here we "adapt" the input using the `.local` call, which somehow "tells to the Kleisli how to pre-compose the simpler input (only the ConfiguredResources) to obtain the composed one by itself.

The resulting `Kleisli` value is a "effectful function" that accepts a `ConfiguredResources` to compute the `IO` effect with no return value (`Unit` is the return type).

Such value of `IO[Unit]` is a "description of how to run a program" that will execute IO effects and eventually return no real value (Unit)